src/lib/storage/buffer/frame.cpp

```diff
@@ -0,0 +1,169 @@
+  return static_cast<NodeID>((state_and_version & NODE_ID_MASK) >> NODE_ID_SHIFT);
+}
+
+bool Frame::try_lock_shared(Frame::StateVersionType old_state_and_version) {
+  auto old_state = state(old_state_and_version);
+  if (old_state < LOCKED_SHARED) {
+    return _state_and_version.compare_exchange_strong(
+        old_state_and_version, update_state_with_same_version(old_state_and_version, old_state + 1));
+  }
+  if (old_state == MARKED) {
+    return _state_and_version.compare_exchange_strong(old_state_and_version,
+                                                      update_state_with_same_version(old_state_and_version, 1));
+  }
+  return false;
+}
+
+bool Frame::try_lock_exclusive(Frame::StateVersionType old_state_and_version) {
+  DebugAssert(state(old_state_and_version) == UNLOCKED || state(old_state_and_version) == MARKED ||
+                  state(old_state_and_version) == EVICTED,
+              "Frame must be unlocked, marked or evicted to lock exclusive, instead: " +
+                  std::to_string(state(old_state_and_version)));
+  return _state_and_version.compare_exchange_strong(old_state_and_version,
+                                                    update_state_with_same_version(old_state_and_version, LOCKED));
+}
+
+bool Frame::unlock_shared() {
+  while (true) {
+    auto old_state_and_version = _state_and_version.load();
+    auto old_state = state(old_state_and_version);
+    Assert(old_state > 0 && old_state <= LOCKED_SHARED, "Frame must be locked shared to unlock shared.");
+    auto new_state = old_state - 1;
+    if (_state_and_version.compare_exchange_strong(old_state_and_version,
+                                                   update_state_with_same_version(old_state_and_version, new_state))) {
+      return new_state == Frame::UNLOCKED;  // return true if last shared lock was released
+    }
+  }
+}
+
+void Frame::unlock_exclusive() {
+  Assert(state(_state_and_version.load()) == LOCKED,
+         "Frame must be locked to unlock exclusive. " + std::to_string(state(_state_and_version.load())));
+  _state_and_version.store(update_state_with_increment_version(_state_and_version.load(), UNLOCKED),
+                           std::memory_order_release);
+}
+
+Frame::StateVersionType Frame::update_state_with_same_version(Frame::StateVersionType old_version_and_state,
+                                                              Frame::StateVersionType new_state) {
+  constexpr auto SHIFT = NUM_BITS - STATE_SHIFT;
+  static_assert(SHIFT == 16, "Shift must be 8.");
+  return ((old_version_and_state << SHIFT) >> SHIFT) | (new_state << STATE_SHIFT);
+}
+
+Frame::StateVersionType Frame::update_state_with_increment_version(Frame::StateVersionType old_version_and_state,
+                                                                   Frame::StateVersionType new_state) {
+  constexpr auto SHIFT = NUM_BITS - STATE_SHIFT;
+  static_assert(SHIFT == 16, "Shift must be 8.");
```

Comment:
When defining `NUM_BITS`, you use `sizeof(StateVersionType) * CHAR_BIT`, `STATE_MASK       = 0xFFFF000000000000`, and `STATE_SHIFT = std::countr_zero(STATE_MASK);`. Since the STATE_MASK is hardcoded and you ensure SHIFT being 16 here, do we have any benefit in not hard-coding `NUM_BITS`?

Reply:
To make things transparent basically.